File: Golang/DataStructures/SequentialStack.go
package DataStructures

import "github.com/JhyeonLee/DataStructuresAndAlgorithms/golang/utils"

type SequentialStack struct {
	elements []int
	top      int
}

func Initialize_SequentialStack() *SequentialStack {
	stack := &SequentialStack{
		top: -1,
	}
	return stack
}

func (S *SequentialStack) Size() int {
	return S.top + 1
}

func (S *SequentialStack) IsEmpty() bool {
	return S.top == -1
}

func (S *SequentialStack) Top() (int, error) {
	if S.IsEmpty() {
		return -1, utils.EmptyStackException()
	}
	return S.elements[S.top], nil
}

func (S *SequentialStack) push(e int) {
	// fullStackException()?

	S.top += 1
	S.elements[S.top] = e
}

func (S *SequentialStack) pop() (int, error) {
	if S.IsEmpty() {
		return -1, utils.EmptyStackException()
	}
	S.top -= 1
	return S.elements[S.top+1], nil
}

File: Python/DataStructures/SequentialStack.py
from utils.utils import emptyStackException

class Sequentialstack():
    def __init__(self):

        self.elements = []
        self.top = -1

    def size(self):
        return self.top + 1
    
    def isEmpty(self):
        return self.top == -1
    
    def top(self):
        if self.isEmpty():
            return emptyStackException
        return self.elements[self.top]
    
    def push(self, e):
        # fullStackException

        self.top += 1
        self.elements[self.top] = e
    
    def pop(self):
        if self.isEmpty():
            return emptyStackException
        self.top -= 1
        return self.elements[self.top + 1]